source/change-streams/change-streams.rst

```diff
@@ -414,25 +462,96 @@ Once a ``ChangeStream`` has encountered a resumable error, it MUST attempt to re
 
 - Perform server selection.
 - Connect to selected server.
-- If the ``ChangeStream`` has not received any changes, and ``resumeAfter`` is not specified, and the max wire version is >= ``7``:
+- If the ``ChangeStream`` has no cached ``documentResumeToken``, and ``resumeAfter`` is not specified, and the max wire version is >= ``7``:
+
+  - The driver MUST execute the known aggregation command.
+  - If the ``ChangeStream`` has a saved ``postBatchResumeToken``:
+
+    - The driver MUST specify ``resumeAfter`` with the cached ``postBatchResumeToken``.
+  - Else:
 
-    - The driver MUST execute the known aggregation command.
     - The driver MUST specify the ``startAtOperationTime`` key set to the ``startAtOperationTime`` provided by the user or saved from the original aggregation.
     - The driver MUST NOT set a ``resumeAfter`` key.
-    - In this case, the ``ChangeStream`` will return all changes that occurred after the specified ``startAtOperationTime``.
 - Else:
 
-    - The driver MUST execute the known aggregation command.
-    - The driver MUST specify a ``resumeAfter`` with the last known ``resumeToken``.
-    - The driver MUST NOT set a ``startAtOperationTime``.
-    - If a ``startAtOperationTime`` key was part of the original aggregation command, the driver MUST remove it.
-    - In this case, the ``ChangeStream`` will return notifications starting with the oplog entry immediately *after* the provided token.
+  - The driver MUST execute the known aggregation command.
```

Comment:
If I'm understanding this list correctly, the `Else` just above this line is for a case where `resumeAfter` was specified by the user. In the lines that follow, it looks like `resumeAfter` is only ever populated with either the cached `postBatchResumeToken` or `documentResumeToken`. Am I missing something, or do the lines below only take effect _after_ the initial aggregate is executed with the user-specified options?

If this is simply the resume process, then does it matter at all if `resumeAfter` was originally specified by the user ("and ``resumeAfter`` is not specified" in the third list item ~12 lines above).

Reply:
Good catch!

The else is for:
change stream has cached documentResumeToken
OR
change stream was specified with resumeAfter
OR
max wire version < 7

But yes, there should be another elseif to account for a resumeAfter specified when no postBatchResumeToken or documentResumeToken have been cached.

Reply:
Actually... I feel like there might be a way to better organize these conditions. I'll take another look in a bit.

Reply:
Ok, hopefully this is clearer (albeit verbose).